public/pages/workflow_detail/workspace/workspace.tsx

```diff
@@ -3,24 +3,161 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import React from 'react';
+import React, { useRef, useContext, useCallback, useEffect } from 'react';
+import ReactFlow, {
+  Controls,
+  Background,
+  useNodesState,
+  useEdgesState,
+  addEdge,
+} from 'reactflow';
 import { useSelector } from 'react-redux';
-import { EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
-import { AppState } from '../../../store';
-import { WorkspaceComponent } from '../workspace_component';
+import { EuiFlexItem, EuiFlexGroup } from '@elastic/eui';
+import { AppState, rfContext } from '../../../store';
+import { convertToReactFlowData } from '../../../../common';
 
-export function Workspace() {
-  const { components } = useSelector((state: AppState) => state.workspace);
+// styling
+import 'reactflow/dist/style.css';
+import './reactflow-styles.scss';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface WorkspaceProps {}
+
+export function Workspace(props: WorkspaceProps) {
+  const reactFlowWrapper = useRef(null);
+  const { reactFlowInstance, setReactFlowInstance } = useContext(rfContext);
+
+  // Fetching workspace state to populate the initial nodes/edges.
+  // Where/how the low-level ReactFlow JSON will be persisted is TBD.
+  // TODO: update when that design is finalized
+  const storedComponents = useSelector(
+    (state: AppState) => state.workspace.components
+  );
+  const { rfNodes, rfEdges } = convertToReactFlowData(storedComponents);
+  const [nodes, setNodes, onNodesChange] = useNodesState(rfNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(rfEdges);
+
+  const onConnect = useCallback(
+    (params) => {
+      setEdges((eds) => addEdge(params, eds));
+    },
+    // TODO: add customized logic to prevent connections based on the node's
+    // allowed inputs. If allowed, update that node state as well with the added
+    // connection details.
+    [setEdges]
+  );
+
+  const onDragOver = useCallback((event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'move';
+  }, []);
+
+  const onDrop = useCallback(
```

Comment:
Might require its own different conversation but for things like credentials that users might not want to type in here on the UI we might need to pre-populate some variable during "on drop" here if we allow some pre-configuration

Reply:
Yeah, we can handle this logic outside of the overall workspace component and treat it at the node/component-level. Flowise for example, has a CredentialInputHandler which is a special handler rendered when a credential input is included in the component's interface.

This will require some more digging to understand and implement a secure way for this part, though. Let me add an issue to track this, thanks for the callout.